[Merged by Bors] - Add AND/OR combinators for run conditions

JoJoJet opened the pull request:
# Objective

Fix #7584.

## Solution

Add an abstraction for creating custom system combinators with minimal boilerplate. Use this to implement AND/OR combinators. Use this to simplify the implementation of `PipeSystem`.

## Example

Feel free to bikeshed on the syntax.

I chose the names `and_then`/`or_else` to emphasize the fact that these short-circuit, while I chose method syntax to empasize that the arguments are *not* treated equally.

```rust
app.add_systems((
    my_system.run_if(resource_exists::<R>().and_then(resource_equals(R(0)))),
    our_system.run_if(resource_exists::<R>().or_else(resource_exists::<S>())),
));
```

---

## Todo

- [ ] Decide on a syntax
- [x] Write docs
- [x] Write tests

## Changelog

+ Added the extension methods `.and_then(...)` and `.or_else(...)` to run conditions, which allows combining run conditions with short-circuiting behavior.
+ Added the trait `Combine`, which can be used with the new `CombinatorSystem` to create system combinators with custom behavior.

JoJoJet commented:
@alice-i-cecile @inodentry I'd appreciate your opinions on the syntax used here :)

alice-i-cecile commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -1,12 +1,31 @@
 /// with [read-only](crate::system::ReadOnlySystemParam) parameters.
-pub trait Condition<Params>: sealed::Condition<Params> {}
+pub trait Condition<Params>: sealed::Condition<Params> {
+    fn and_then<P, C: Condition<P>>(self, and_then: C) -> AndThen<Self::System, C::System> {
```
Doc strings and doc tests here :)

I really like that these are methods though!

alice-i-cecile commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -142,3 +161,69 @@ pub mod common_conditions {
+impl<In, A, B> Combine<A, B> for AndThenMarker
+where
+    In: Copy,
+    A: System<In = In, Out = bool>,
```
Should this be restricted to ReadOnlySystem?

JoJoJet replied:
Not necessary, `CombinatorSystem<A, B, AndThenMarker>` only implements `ReadOnlySystem` if `A` and `B` do, so the trait bounds "just work" as is.

alice-i-cecile commented on crates/bevy_ecs/src/system/combinator.rs:
```diff
@@ -0,0 +1,147 @@
+
+use super::{ReadOnlySystem, System};
+
+pub trait Combine<A: System, B: System> {
```
Doc strings.

alice-i-cecile commented on crates/bevy_ecs/src/system/combinator.rs:
```diff
@@ -0,0 +1,147 @@
+    }
+}
+
+/// SAFETY: Both systems are read-only, so the system combining them will only read from the world.
```
```suggestion
/// SAFETY: Both systems are read-only, so any system created by combining them will only read from the world.
```

alice-i-cecile commented:
Milestone placement is "we should make a call on this", not "this must be included". I think there's a lot of value here, but it's also a little tricky.

DJMcNab commented:
Why doesn't `pipe` use this combinator pattern?

I'd like us to avoid proliferating unsafe system implementations.

Two? is already too many, can we avoid it growing to three?

JoJoJet commented:
> Why doesn't pipe use this combinator pattern?

It does.

DJMcNab commented:
Ah, the dangers of trying to do code review on my phone. My apologies.

DJMcNab reviewed (commented):
I'm not convinced by the safety comment on the calls to run_unsafe. They seem a bit blank.

(I also believe that they're not sound)

I can't say I especially like us stacking more content into `System` impls rather than just using proper systems. However, until Rust catches up, that's inevitable.

I also believe this new system is feasibly isomorphic to `ChainSystem`, but I also think the ergonomics are probably better.

JoJoJet commented:
> (I also believe that they're not sound)

It's unsound in the same way that every other `System` impl is (has no guarantees that it's running on the correct `World`). I've been meaning to open a PR to address this.

JoJoJet commented on crates/bevy_ecs/src/system/combinator.rs:
```diff
@@ -0,0 +1,228 @@
+    type In;
+    type Out;
+
+    fn combine(
```
I ended up rewriting this trait and the way it's used, so this suggestion doesn't apply anymore. I also added comments for the relevant safety invariants, which should hopefully be more convincing now.

JoJoJet commented:
I've added documentation and examples, simplified the `Combine` trait, and documented the previously-implicit safety invariants in `impl System for CombinatorSystem`.

alice-i-cecile commented on crates/bevy_ecs/src/system/combinator.rs:
```diff
@@ -0,0 +1,228 @@
+
+/// A [`System`] defined by combining two other systems.
+/// The behavior of this combinator is specified by implementing the [`Combine`] trait.
+/// For a full usage example, see the docs for `Combine`.
```
```suggestion
/// For a full usage example, see the docs for [`Combine`].
```

alice-i-cecile reviewed (approved):
Very nice to see this.

Shatur commented:
> I chose the names and_then/or_else to emphasize the fact that these short-circuit, while I chose method syntax to empasize that the arguments are not treated equally.

Maybe use `and` and `or`? Shorter and a little more readable as to me.

JoJoJet commented:
Personally I prefer `and_then`/`or_else` for the reasons I mentioned in the PR desc -- I think that making the behavior clear is more important than making the name short. I'm willing to change it if there's a consensus for those names (or any other) though, of course.

alice-i-cecile commented:
I prefer `and_then`/`or_else` as well :)

BlackPhlox commented:
Could you add an example of usage under `examples/ecs`? There might already be a suitable .rs file

JoJoJet commented:
From what I can tell, none of our examples use run conditions at all (except for `.on_update()`). We should address this, but IMO there should be a separate issue for this.

Shatur commented:
Will it be possible to combine multiple conditions with this syntax?..

For example something like `A && B || C && D`?
I think it's very useful if you have substaces. For example, in my game I run `PlacingObjectPlugin::movement_system` in `GameState::City` with substate `CityMode::Objects` **and** in `GameState::Family` with substate `FamilyMode::Building`. Currently I have to add my system twice ([1](https://github.com/dollisgame/dollis/blob/98ddab5d74d25109f9b022e4d0a14381ffba1bc9/src/core/object/placing_object.rs#L49-L52), [2](https://github.com/dollisgame/dollis/blob/98ddab5d74d25109f9b022e4d0a14381ffba1bc9/src/core/object/placing_object.rs#L94-L97), the game currenly uses loopless, but will be similar with stageless).

JoJoJet commented:
Yes, that should be possible. `and_then` and `or_else` return run conditions, so you should be able to keep adding as many combinators as you want.

alice-i-cecile commented:
Can you add these to the example from #7652 once it's merged?

JoJoJet commented:
I've added `and_then` to the example, but I'm really struggling to think of a way to show off `or_else` that isn't extremely contrived.

alice-i-cecile reviewed (approved):
Yeah, let's go with this: it'll hint to users that they should ask about the OR variant too.

inodentry commented:
The most common use case for OR-combining run conditions that I've seen is to make a system run in multiple states (or other state-like resource type).

For example:

```rust
enum GameMode {
    Singleplayer,
    Multiplayer,
    Tutorial,
}

app.add_system(move_player_units.run_if(
    resource_equals(GameMode::Singleplayer)
        .or_else(resource_equals(GameMode::Multiplayer))
));
app.add_system(move_ai_units.run_if(
    resource_equals(GameMode::Singleplayer)
        .or_else(resource_equals(GameMode::Tutorial))
));
```

You could come up with something along those lines, for the example. It does not seem "contrived" to me. Very much a real-world use-case.

inodentry reviewed (approved):
I like this. I like the syntax, and also the implementation. Read through the code, didn't notice anything wrong. The Combine trait is clever, but useful. :)

The documentation is adequate.
